### PR TITLE
Convert toData to async [API-511]

### PR DESCRIPTION
**Pull request by sancar**

### CODE REVIEW

File: hazelcast/include/hazelcast/client/imap.h [hunk 1]
```diff
@@ -120,8 +152,21 @@ namespace hazelcast {
...
+                return to_data2(value).then(boost::launch::sync,[this, data1, ttl](boost::future<serialization::pimpl::data> f) {
+                  serialization::pimpl::data data2 = f.get();
+                  return put_internal(data1, data2, ttl).then(boost::launch::sync, [this](boost::future<boost::optional<serialization::pimpl::data>> f) {
+                    boost::optional<serialization::pimpl::data> responseData = f.get();
+                    return to_object2<R>(responseData).then(boost::launch::sync, [this](boost::future<boost::optional<R>> f){
+                      return f.get();
+                    });
+                  }).unwrap();
+                }).unwrap();
+              }).unwrap();
```
Comment (yemreinci) [inline hazelcast/include/hazelcast/client/imap.h:169]: We can use the `.then().unwrap().then().unwrap()...` pattern to flatten the callback hell.

Both of the following alternatives compiled and passed tests for me:

=== using when_all ===

```suggestion
              using boost::launch;
              using serialization::pimpl::data;

              std::vector<boost::future<data>> data_futures;
              data_futures.push_back(to_data2(key));
              data_futures.push_back(to_data2(value));

              auto result =
                  boost::when_all(data_futures.begin(), data_futures.end())
                      .then(
                          launch::sync,
                          [this, ttl](boost::future<boost::csbl::vector<boost::future<data>>> f) {
                            auto data_vector = f.get();
                            auto key_data = data_vector[0].get();
                            auto value_data = data_vector[1].get();
                            return put_internal(key_data, value_data, ttl);
                          })
                      .unwrap()
                      .then(launch::sync,
                            [this](boost::future<boost::optional<data>> f) {
                              auto result = f.get();
                              return to_object2<R>(result);
                            })
                      .unwrap();

              return result;
```

====== without when_all ====
```suggestion
              using boost::launch;
              using serialization::pimpl::data;

              struct state_ {
                boost::future<data> key_data_fut;
                boost::future<data> value_data_fut;
              };

              auto state = std::make_shared<state_>();
              state->key_data_fut = to_data2(key);
              state->value_data_fut = to_data2(value);

              auto result = state->key_data_fut
                  .then(launch::sync, [state](boost::future<data> f){
                    // store the ready key_data_fut for later
                    state->key_data_fut = std::move(f);
                    return std::move(state->value_data_fut);
                  })
                  .unwrap()
                  .then(launch::sync, [state, ttl, this](boost::future<data> f){
                    auto key_data = state->key_data_fut.get();
                    auto value_data = f.get();
                    return put_internal(key_data, value_data, ttl);
                  })
                  .unwrap()
                  .then(launch::sync, [this](boost::future<boost::optional<data>> f){
                    return to_object2<R>(f.get());
                  })
                  .unwrap();

              return result;
```

File: hazelcast/include/hazelcast/client/imap.h [hunk 2]
```diff
@@ -120,8 +128,36 @@ namespace hazelcast {
...
             template<typename K, typename V, typename R=V>
-            boost::future<boost::optional<R>> put(const K &key, const V &value, std::chrono::milliseconds ttl) {
-                return to_object<R>(put_internal(to_data(key), to_data(value), ttl));
+            boost::future<boost::optional<R>>
+            put(const K &key, const V &value, std::chrono::milliseconds ttl) {
+                using boost::launch;
+                using boost::future;
+                using serialization::pimpl::data;
+                struct state_ {
+                  future<data> key_data_fut;
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/imap.h:137]: Did you consider using `boost::when_all` or `boost::when_any`. E.g.:
```
            template<typename K, typename V, typename R=V>
            boost::future<boost::optional<R>>
            put(const K &key, const V &value, std::chrono::milliseconds ttl) {
                using boost::launch;
                using boost::future;
                using serialization::pimpl::data;
                return boost::when_all(to_data2(key), to_data2(value)).then(sync, [this, ttl] (std::vector<future<data>> futures) {
                    return std::make_pair(futures[0].get(), futures[1].get());
                }).then(sync, [this, ttl] (future<std::pair<data, data>> f){
                    auto data_pair  = f.get();
                    return put_internal(data_pair.first, data_pair.second, ttl);
                }).unwrap();
            }
```

I still don't want to use method `unwrap` which is experimental, there may be a way to even eliminate that.

Comment (sancar) [reply hazelcast/include/hazelcast/client/imap.h:137]: when_all opens a thread that is why I avoided it.

File: hazelcast/include/hazelcast/client/serialization/serialization.h [hunk 3]
```diff
@@ -1308,33 +1309,41 @@ namespace hazelcast {
                     DataSerializer &get_data_serializer();
 
                     template<typename T>
-                    inline data to_data(const T *object) {
-                        object_data_output output(serialization_config_.get_byte_order(), false, &portable_serializer_, serialization_config_.get_global_serializer());
+                    inline boost::future<data> to_data2(const T *object) {
```
[the remark this answers is not in the document]
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/serialization/serialization.h:1312]: I think Sancar will rename to_data2 with to_data and remove the old to_data when he completes his work. The bad naming (to_data2) is just temporary for now until he gets things working.

File: hazelcast/include/hazelcast/client/imap.h (hunk 2 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/imap.h:139]: > I still don't want to use method unwrap which is experimental, there may be a way to even eliminate that.

The `boost::future::then` is also experimental: https://www.boost.org/doc/libs/1_78_0/doc/html/thread/synchronization.html#thread.synchronization.futures.reference.unique_future.then

**Comment by sancar** [comment]
We have run performance benchmarks against 3 branches. master (https://github.com/hazelcast/hazelcast-cpp-client/commit/ab322dd7b18b0058d99bb528cf81d85b77064006), https://github.com/sancar/hazelcast-cpp-client/tree/asyncToData and https://github.com/sancar/hazelcast-cpp-client/tree/toObjectSyncPolicy 

The result is that converting toData and toObject to async causes 14/15% performance penalty just because indirections(there are not remote calls) 

Raw results:
```
# 1 Client 1 Thread 2 Nodes 
## master
map_put/threads:1_mean       784127 ns       113661 ns           10 Put Rate=8.80297k/s
map_put/threads:1_median     790161 ns       114311 ns           10 Put Rate=8.74809k/s
map_put/threads:1_stddev      24135 ns         2755 ns           10 Put Rate=222.843/s
## toObjectSyncPolicy
map_put/threads:1_mean       656075 ns       103389 ns           10 Put Rate=9.82257k/s
map_put/threads:1_median     658942 ns       106867 ns           10 Put Rate=9.35758k/s
map_put/threads:1_stddev      21660 ns        11524 ns           10 Put Rate=1.50171k/s
## asyncToData
map_put/threads:1_mean       750338 ns       157633 ns           10 Put Rate=6.34437k/s
map_put/threads:1_median     749162 ns       157815 ns           10 Put Rate=6.33655k/s
map_put/threads:1_stddev       8474 ns         1499 ns           10 Put Rate=59.9557/s
# 1 Client 1 Thread 2 Nodes
## master
map_put/threads:1_mean       674592 ns        92246 ns           10 Put Rate=10.9672k/s
map_put/threads:1_median     689285 ns        95726 ns           10 Put Rate=10.4476k/s
map_put/threads:1_stddev      28770 ns         9307 ns           10 Put Rate=1.40052k/s
## toObjectSyncPolicy
map_put/threads:1_mean       660656 ns        96008 ns           10 Put Rate=10.4175k/s
map_put/threads:1_median     662841 ns        96009 ns           10 Put Rate=10.4157k/s
map_put/threads:1_stddev      11681 ns         1304 ns           10 Put Rate=141.553/s
## asyncToData
map_put/threads:1_mean       698711 ns       136701 ns           10 Put Rate=7.31839k/s
map_put/threads:1_median     701106 ns       136289 ns           10 Put Rate=7.33754k/s
map_put/threads:1_stddev      15638 ns         3012 ns           10 Put Rate=158.095/s
# 1 Client 1 Thread 2 Nodes 
## master
map_put/threads:1_mean       639413 ns        89007 ns           10 Put Rate=11.4356k/s
map_put/threads:1_median     654174 ns        94174 ns           10 Put Rate=10.6187k/s
map_put/threads:1_stddev      38117 ns        11221 ns           10 Put Rate=1.7693k/s
## toObjectSyncPolicy
map_put/threads:1_mean       496257 ns        60682 ns           10 Put Rate=19.4792k/s
map_put/threads:1_median     504718 ns        67677 ns           10 Put Rate=14.7867k/s
map_put/threads:1_stddev      38712 ns        16378 ns           10 Put Rate=12.8315k/s
## asyncToData
map_put/threads:1_mean       700208 ns       137646 ns           10 Put Rate=7.26537k/s
map_put/threads:1_median     700591 ns       137564 ns           10 Put Rate=7.26933k/s
map_put/threads:1_stddev       6775 ns         1021 ns           10 Put Rate=53.5194/s
# 1 Client 12 Threads 2 Nodes  
## master
map_put/threads:12_mean        28066 ns        22796 ns           10 Put Rate=43.8744k/s
map_put/threads:12_median      27952 ns        22775 ns           10 Put Rate=43.9087k/s
map_put/threads:12_stddev        945 ns          312 ns           10 Put Rate=596.243/s
## toObjectSyncPolicy
map_put/threads:12_mean        25478 ns        23774 ns           10 Put Rate=42.1167k/s
map_put/threads:12_median      25432 ns        23604 ns           10 Put Rate=42.3653k/s
map_put/threads:12_stddev        681 ns          893 ns           10 Put Rate=1.57284k/s
## asyncToData
map_put/threads:12_mean        27053 ns        56328 ns           10 Put Rate=17.7614k/s
map_put/threads:12_median      26716 ns        56723 ns           10 Put Rate=17.6295k/s
map_put/threads:12_stddev        860 ns         1266 ns           10 Put Rate=406.999/s
# 1 Client 12 Threads 2 Nodes  
## master
map_put/threads:12_mean        28373 ns        22630 ns           10 Put Rate=44.1975k/s
map_put/threads:12_median      27805 ns        22634 ns           10 Put Rate=44.1812k/s
map_put/threads:12_stddev       1613 ns          341 ns           10 Put Rate=667.31/s
## toObjectSyncPolicy
map_put/threads:12_mean        24490 ns        24487 ns           10 Put Rate=40.8881k/s
map_put/threads:12_median      24549 ns        24405 ns           10 Put Rate=40.9753k/s
map_put/threads:12_stddev        503 ns          900 ns           10 Put Rate=1.52788k/s
## asyncToData
map_put/threads:12_mean        27442 ns        55600 ns           10 Put Rate=17.9929k/s
map_put/threads:12_median      27875 ns        55637 ns           10 Put Rate=17.9736k/s
map_put/threads:12_stddev        823 ns         1166 ns           10 Put Rate=379.882/s
```

**Pull request closed without merging**